src/gate_detect/src/GateDetection.cpp

```diff
@@ -0,0 +1,330 @@
+    cv::GaussianBlur(matin, gaussianBlurred, cv::Size(7, 7), 0, 0);
+
+    cv::medianBlur(gaussianBlurred, medianBlurred, 7);
+
+    cv::Mat dst;
+
+    cv::Canny(medianBlurred, dst, _cannyLow, _cannyLow * 3, 3);
+
+    std::vector<cv::Vec4i> detectedLines;
+
+    cv::HoughLinesP(dst,
+                    detectedLines,
+                    1,
+                    CV_PI / 180,
+                    _houghLinesThreshold,
+                    _houghLinesMinLength,
+                    _houghLinesMaxLineGap);
+
+    std::vector<float> gateVector = std::vector<float>(9, 0);
+
+    if (detectedLines.empty()) return gateVector;
+
+    std::vector<cv::Vec4i> vertLines = filterVertLines(detectedLines);
+
+    std::vector<PoleDetection> vertPoles = findVertPoles(vertLines);
+
+    std::vector<cv::Vec4i> horLines = filterHorLines(detectedLines);
+
+    std::vector<PoleDetection> horPoles = findHorPoles(horLines);
+
+    gateVector = getGateVector(vertPoles, horPoles);
+
+    return gateVector;
+}
+
+std::vector<cv::Vec4i>
+GateDetection::filterVertLines(std::vector<cv::Vec4i> detectedLines) {
+    std::vector<cv::Vec4i> vertLines;
+
+    std::copy_if(detectedLines.begin(),
+                 detectedLines.end(),
+                 std::back_inserter(vertLines),
+                 [=](cv::Vec4i line1) {
+                     return abs((line1)[0] - (line1)[2]) < _lowVertThresh;
+                 });
+
+    return vertLines;
+}
+
+std::vector<cv::Vec4i>
+GateDetection::filterHorLines(std::vector<cv::Vec4i> detectedLines) {
+    std::vector<cv::Vec4i> horLines;
+
+    std::copy_if(detectedLines.begin(),
+                 detectedLines.end(),
+                 std::back_inserter(horLines),
+                 [=](cv::Vec4i line1) {
+                     return abs((line1)[1] - (line1)[3]) < _lowHorThresh;
+                 });
+
+    return horLines;
+}
+
+std::vector<PoleDetection>
+GateDetection::findVertPoles(std::vector<cv::Vec4i> vertLines) {
+    std::vector<PoleDetection> verticalPoles;
+
+    std::vector<cv::Vec4i>::iterator line1;
+    std::vector<cv::Vec4i>::iterator line2;
+
+    if (vertLines.size() <= 1)
+        return verticalPoles;
+
+    else if (vertLines.size() == 2) {
+        verticalPoles.push_back(
+        PoleDetection(vertLines.at(0), vertLines.at(1), m_Hor, b_Hor));
+
+        return verticalPoles;
+    }
+
+    for (line1 = vertLines.begin(); line1 != vertLines.end(); line1++) {
+        for (line2 = std::next(line1); line2 != vertLines.end();) {
+            if ((abs((*line1)[0] - (*line2)[0]) < 2) || (*line1 == *line2)) {
+                line2 = vertLines.erase(line2);
+
+            } else if (abs((*line1)[0] - (*line2)[0]) < _poleMax) {
+                PoleDetection verticalPole =
+                PoleDetection(*line1, *line2, m_Vert, b_Vert);
+
+                verticalPoles.push_back(verticalPole);
+
+                line2 = vertLines.erase(line2);
+
+            } else {
+                line2++;
+            }
+        }
+    }
```

Comment:
We should add a bunch of comments to this function explaining what we're doing at each step. It would also be helpful if for complex conditionals we split them out. For example `dx_start = abs((*line1)[0] - (*line2)[0])`

Reply:
defined comments for algorithm